include/uam/io/tcp_client.h

```diff
@@ -0,0 +1,294 @@
+#include <boost/make_shared.hpp>
+#include <boost/system/error_code.hpp>
+
+#include <ros/init.h>
+#include <uam/shape_shifter_packet.h>
+#include <uam/uam_visitors.h>
+
+#include <memory>
+#include <string>
+
+namespace uam
+{
+/**
+ * @brief TCP client class using boost::asio
+ */
+class TcpClient
+{
+public:
+  /**
+   * @brief type of the Callback that receives the packet
+   */
+  using OnNewDataCallback = std::function<void(const protocol::ShapeShifterPacket& packet, const ros::Time&)>;
+
+  /**
+   * @brief This
+   */
+  using FilterCallback = std::function<bool(const decltype(protocol::CommandReplyHeader::header)& packet)>;
+
+  /**
+   * @brief Default C'tor
+   */
+  explicit TcpClient(OnNewDataCallback callback);
+
+  /**
+   * @brief
+   *
+   * @param name
+   * @param callback
+   * @param external_context
+   */
+  TcpClient(OnNewDataCallback callback, std::shared_ptr<boost::asio::io_service>& external_context);
+
+  /**
+   * @brief Default D'tor
+   */
+  ~TcpClient();
+
+  /**
+   * @brief Connect to tcp server
+   *
+   * @param[in] remote_ip - Server ip address
+   * @param[in] remote_port - Server port
+   *
+   * @return true if connect successful, false otherwise
+   */
+  bool connect(const std::string& remote_ip = "", uint16_t remote_port = 0u);
+
+  /**
+   * @brief Returns the state of the socket
+   *
+   * @return true if connected, false otherwise
+   */
+  inline bool isConnected() const { return connected_; }
+
+  /**
+   * @brief Returns if async read is queued or not
+   *
+   * @return true if async read is in progress, alse otherwise
+   */
+  inline bool isStopped() const { return stopped_; }
+
+  /**
+   * @brief Disconnect client
+   */
+  void disconnect();
+
+  /**
+   * @brief Place and send order and wait for it to be sent
+   *
+   * @param[in] command - Command to send
+   * @return true if command successfully sent, false otherwise.
+   */
+  inline bool asyncSend(const std::string& command)
+  {
+    if (!connected_)
+      return false;
+    auto send_length = socket_.async_send(boost::asio::buffer(command, command.size()), boost::asio::use_future);
+    return (send_length.get() == command.size());
+  }
+
+  /**
+   * @brief Start Async Read Task (to read UAM commands)
+   *
+   * @return true if async task was created false otherwise
+   */
+  inline bool startAsyncReadTask()
+  {
+    if (!connected_)
+    {
+      ROS_WARN_STREAM("Not connected, cannot start async read task");
+      return false;
+    }
+
+    if (!stopped_)
+    {
+      return true;
+    }
+
+    asyncReadData();
+    stopped_ = false;
+    stop_requested_ = false;
+    return true;
+  }
+
+  /**
+   * @brief Create an async read line operation
+   *
+   * @param[in] handler - Async read handler
+   */
+  template <typename TReply, typename Handler>
+  void asyncReadLine(Handler handler)
+  {
+    if (!stopped_)
+    {
+      ROS_ERROR_STREAM("Cannot send async readline with async receive in progress.");
```

Comment:
Sorry, I am not following. could you elaborate?

Reply:
Oh, it's been way too long to expect me to remember.

I guess what I am asking:
In the `asyncReadLine()` function, you guard against starting a "read line" operation if there is a "read task" operation already in progress.
However, the "read line" operation is also an async operation, which should return from the call before the "read line" completes. Do you need to add similar guard behavior so you do not start a "read task" while a "read line" is pending?

Reply:
> Oh, it's been way too long to expect me to remember.

Sorry for that. 
From the driver's perspective, that will not happen, but from this class's perspective, that can definitely happen. I will add a todo to add that guard (if that is acceptable to you). Otherwise I might need a bit of refactor to make this look good and not just another boolean flagging that async operation.

Reply:
A TODO works for me at the moment.